### PR TITLE
Add vtu output capability to TetmeshFacetedSurface

### PR DESCRIPTION
**Pull request by RupinderM**
### Feature
The functions output_paraview(std::ofstream& outfile) and output_paraview(const std::string filename) have been added as public member functions of TetmeshFacetedSurface.

The function output_paraview(std::ofstream& outfile) contains the code that creates the .vtu file while the other version simply creates an ofstream to a file with the provided filename and calls output_paraview(std::ofstream& outfile) again.

These functions create a .vtu file from the object that the method is called on, making it easy to view TetmeshFacetedSurfaces in Paraview.

```cpp
class NewFacetedSurface: public TetMeshFacetedSurface
{
... Relevant code defining the faceted surface...
};

int main()
{
NewFacetedSurface faceted_surface(<constructor arguments>);

// To call the version of output_paraview that takes a std::string argument
faceted_surface.output_paraview(<filename with .vtu extension>);

// Or the following if you wish to call the version with the std::ofstream argument
std::ofstream outfile(<filename with .vtu extension>);
faceted_surface.output_paraview(outfile);
outfile.close();
}
```

### Self-Test
The new self test for this capability is in self_test/faceted_surface_vtu_output.
Several faceted surfaces defined in the demo drivers have been instantiated in the driver code faceted_surface_vtu_output_tester.cc. Then output_paraview has been called on all of these faceted surfaces. The resulting .vtu files may be viewed in Paraview.

There is a validate.sh script that simply executes the driver code in faceted_surface_vtu_output, if it executes without fail then validate.sh will pass.

To run the code, simply run the following shell commands from /path/to/oomph-lib/self_test/faceted_surface_vtu_output

```bash
make faceted_surface_vtu_output_tester && ./faceted_surface_vtu_output_tester
```

### CODE REVIEW

File: self_test/faceted_surface_vtu_output/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+endif()
+include(CTest)
+
+oomph_add_executable(
+  NAME faceted_surface_vtu_output_tester
+  SOURCES faceted_surface_vtu_output_tester.cc
+  LIBRARIES oomph::poisson oomph::generic)
+
+message(VERBOSE "Leaving faceted_surface_vtu_output subdirectory")
+# ------------------------------------------------------------------------------
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/CMakeLists.txt:18]: I think you've mistakenly added this file; the main `oomph-lib` repository doesn't use CMake files (yet!)

File: self_test/faceted_surface_vtu_output/spherical_tetmesh_faceted_surface.h [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+      Facet_pt[19] = new TetMeshFacet(n_vertex_on_facet);
+      Facet_pt[19]->set_vertex_pt(0, Vertex_pt[2]);
+      Facet_pt[19]->set_vertex_pt(1, Vertex_pt[3]);
+      Facet_pt[19]->set_vertex_pt(2, Vertex_pt[11]);
+
+      // icosa_facet[19].resize(3);
+      // icosa_facet[19][0] = 2;
+      // icosa_facet[19][1] = 3;
+      // icosa_facet[19][2] = 11;
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/spherical_tetmesh_faceted_surface.h:289]: If these comment blocks are not needed, please remove them.

Comment (puneetmatharu) [reply self_test/faceted_surface_vtu_output/spherical_tetmesh_faceted_surface.h:289]: (This includes the other comment blocks above.)

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 3]
```diff
@@ -0,0 +1,1256 @@
...
+#endif
+
+
+namespace oomph
+{
+
+//================================================================
+/// Helper function
+//================================================================
+namespace HelperForSortingVectorOfPairsOfVertexPtAndDouble
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:44]: Nice and verbose - I like it!

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 4]
```diff
@@ -0,0 +1,1256 @@
...
+   
+   box_point[0] = offset[0]+box_half_width;
+   box_point[1] = offset[1]+box_half_width;
+   box_point[2] = offset[2]+box_half_length;
+   Vertex_pt[6]=new TetMeshVertex(box_point);
+   
+   box_point[0] = offset[0]+box_half_width;
+   box_point[1] = offset[1]-box_half_width;
+   box_point[2] = offset[2]+box_half_length;
+   Vertex_pt[7]=new TetMeshVertex(box_point);
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:191]: Has this code been formatted by `clang-format`? It's odd to see no spaces around the `+`/`-` operators... Especially given that there _is_ a space around the `=` operator (for most cases). Please check that the code has been formatted; if not, please add spaces around the operators for consistency.

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 5]
```diff
@@ -0,0 +1,1256 @@
...
+
+   // Make facets
+   unsigned n_facet=6;
+   Facet_pt.resize(n_facet);
+   
+   unsigned n_vertex_on_facet=4;
+   Facet_pt[0]=new TetMeshFacet(n_vertex_on_facet);
+   unsigned one_based_boundary_id=1+one_based_boundary_id_offset;
+   Facet_pt[0]->set_one_based_boundary_id(one_based_boundary_id);   
+   Facet_pt[0]->set_vertex_pt(0,Vertex_pt[0]);
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:202]: Please add spaces around the operators and after the comma in `set_vertex_pt(0,Vertex_pt[0]);`

Comment (puneetmatharu) [reply self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:202]: ...and for the code below

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 6]
```diff
@@ -0,0 +1,1256 @@
...
+   Facet_pt[5]->set_one_based_boundary_id(one_based_boundary_id);   
+   Facet_pt[5]->set_vertex_pt(0,Vertex_pt[5]);
+   Facet_pt[5]->set_vertex_pt(1,Vertex_pt[1]);
+   Facet_pt[5]->set_vertex_pt(2,Vertex_pt[0]);
+   Facet_pt[5]->set_vertex_pt(3,Vertex_pt[4]);
+
+
+   // Specify the adjacent region that is bounded by these facets
+   for (unsigned f=0;f<6;f++)
+    {
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:252]: Hmm, more formatting oddities; the braces don't seem to align with the `f` on the preceding line and no spaces around operators...

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 7]
```diff
@@ -0,0 +1,1256 @@
...
+   Facet_pt[4]->set_vertex_pt(3,Vertex_pt[0]);
+
+   // bottom
+   Facet_pt[5]=new TetMeshFacet(n_vertex_on_facet);
+   one_based_boundary_id=6+one_based_boundary_id_offset;
+   Facet_pt[5]->set_one_based_boundary_id(one_based_boundary_id);   
+   Facet_pt[5]->set_vertex_pt(0,Vertex_pt[5]);
+   Facet_pt[5]->set_vertex_pt(1,Vertex_pt[1]);
+   Facet_pt[5]->set_vertex_pt(2,Vertex_pt[0]);
+   Facet_pt[5]->set_vertex_pt(3,Vertex_pt[4]);
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:247]: I'm not going to go through these indices by hand and check that they are correct -- I'm just going to assume that you're right! However, @MatthiasHeilManchester may check these if he's bored...

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 8]
```diff
@@ -0,0 +1,1256 @@
...
+
+
+private:
+
+ /// Build the thing
+ void build_it(const double& half_x_width, 
+               const double& half_y_length,
+               const Vector<double>& offset,
+               const unsigned& one_based_boundary_id)
+  {
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:312]: Any reason why this code doesn't just go inside the constructor?

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 9]
```diff
@@ -0,0 +1,1256 @@
...
+   // Make vertices
+   unsigned n_vertex=4;
+   Vertex_pt.resize(n_vertex);
+   Vector<double> box_point(3);
+
+   box_point[0] = offset[0]-half_x_width;
+   box_point[1] = offset[1]-half_y_length;
+   box_point[2] = offset[2];
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:320]: Formatting

File: src/generic/tet_mesh.h [hunk 10]
```diff
@@ -408,6 +408,192 @@ namespace oomph
...
+      // Vertices
+      //---------
+
+      outfile << "<Points>\n"
+              << "<DataArray type=\"Float32\" NumberOfComponents=\"3\" "
+                 "format=\"ascii\">"
+              << std::endl;
+
+      // Output the coordinates of every distinct vertex
+      for (auto vertex : vertices)
```
Comment (puneetmatharu) [inline src/generic/tet_mesh.h:521]: Make these for-range loops use const references where possible, e.g.

```cpp
for (const auto& vertex : vertices)
...
```

File: self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h [hunk 11]
```diff
@@ -0,0 +1,1256 @@
...
+        }
+      }
+    }
+
+   //Find mean position
+   double z_mean=0.0;
+   double z_min= DBL_MAX;
+   double z_max=-DBL_MAX;
+   unsigned n=sorted_perimeter_vertex_pt.size();
+   for (unsigned j=0;j<n;j++)
```
Comment (puneetmatharu) [inline self_test/faceted_surface_vtu_output/tetmesh_faceted_surfaces.h:1157]: Could use a for-range loop to simplify this

**Review by puneetmatharu — CHANGES_REQUESTED** [review]
Looks good to me for the most part apart from the fact that you need to run `clang-format` over these files (or if you have, find out why the formatting is so odd). 

Note that I have assumed that you have done all of the maths/setting up indices correctly and haven't checked them myself. @MatthiasHeilManchester do you want to look through the code too?

File: self_test/faceted_surface_vtu_output/CMakeLists.txt (hunk 1 above)
Comment (RupinderM) [reply self_test/faceted_surface_vtu_output/CMakeLists.txt]: I've added a commit (https://github.com/oomph-lib/oomph-lib/pull/98/commits/d3231c530db11245c6ce921654fb0f9d44117ed6) that removes this CMakeLists.txt file. I can rebase and completely remove this mistake from the history if you prefer.

File: self_test/faceted_surface_vtu_output/spherical_tetmesh_faceted_surface.h (hunk 2 above)
Comment (RupinderM) [reply self_test/faceted_surface_vtu_output/spherical_tetmesh_faceted_surface.h]: These have been removed in a new commit: https://github.com/oomph-lib/oomph-lib/pull/98/commits/6febfdfeb9c794f7880b4b02a0e087c25b1994d5

File: src/generic/tet_mesh.h (hunk 10 above)
Comment (RupinderM) [reply src/generic/tet_mesh.h]: Fixed these issues in another commit: https://github.com/oomph-lib/oomph-lib/pull/98/commits/d93f2d9a9066f2626ab1076da0cc5dc579970122

**Comment by MatthiasHeilManchester** [comment]
Agree with Puneet's comment re the odd formatting. If clangformat tidies this up automatically post-merge then ok (though shouldn't this have run already when you pushed it to your own forked repo on github?).

**Comment by puneetmatharu** [comment]
Merging this in as it looks good now

> (though shouldn't this have run already when you pushed it to your own forked repo on github?).

The clang-format workflow only formats code in the `src/` directory so the `self_test/` directory would be unaffected unfortunately. We should probably update the workflow to ensure that all _new_ C++ files to the repo get formatted by `clang-format`...

**Review by puneetmatharu — APPROVED** [review]

**Pull request merged**